browser/ui/views/brave_actions/brave_shields_action_view.cc

```diff
@@ -100,6 +85,59 @@ SkPath BraveShieldsActionView::GetHighlightPath() const {
   return path;
 }
 
+std::unique_ptr<IconWithBadgeImageSource>
+BraveShieldsActionView::GetImageSource() {
+  auto preferred_size = GetPreferredSize();
+  std::unique_ptr<IconWithBadgeImageSource> image_source(
+      new BraveActionIconWithBadgeImageSource(preferred_size));
+  auto* web_contents = tab_strip_model_->GetActiveWebContents();
+  std::unique_ptr<IconWithBadgeImageSource::Badge> badge;
+  bool is_enabled = false;
+  std::string badge_text;
+
+  if (web_contents) {
+    auto* shields_data_ctrlr =
+        brave_shields::BraveShieldsDataController::FromWebContents(
+            web_contents);
+    int count = shields_data_ctrlr->GetTotalBlockedCount();
+
+    badge_text = count > 0 ? std::to_string(count) : std::string();
+    is_enabled = shields_data_ctrlr->GetIsBraveShieldsEnabled();
+
+    if (!badge_text.empty()) {
+      badge = std::make_unique<IconWithBadgeImageSource::Badge>(
+          badge_text, SK_ColorWHITE, kBadgeBg);
+    }
+  }
+
+  image_source->SetIcon(gfx::Image(GetIconImage(is_enabled)));
+  image_source->SetBadge(std::move(badge));
+
+  return image_source;
+}
+
+gfx::ImageSkia BraveShieldsActionView::GetIconImage(bool is_enabled) {
+  ui::ResourceBundle& rb = ui::ResourceBundle::GetSharedInstance();
+  gfx::ImageSkia image;
+  const SkBitmap bitmap =
+      rb.GetImageNamed(is_enabled ? IDR_BRAVE_SHIELDS_ICON_64
+                                  : IDR_BRAVE_SHIELDS_ICON_64_DISABLED)
+          .AsBitmap();
+  float scale = static_cast<float>(bitmap.width()) / kBraveActionGraphicSize;
+  image.AddRepresentation(gfx::ImageSkiaRep(bitmap, scale));
+  return image;
+}
+
+void BraveShieldsActionView::UpdateIconState() {
+  auto preferred_size = GetPreferredSize();
+  gfx::ImageSkia icon(
+      gfx::Image(gfx::ImageSkia(GetImageSource(), preferred_size))
+          .AsImageSkia());
```

Comment:
Above could be `const gfx::ImageSkia icon(GetImageSource(), GetPreferredSize());`